Added valueset, QR code generation, test-data parser and some fixes

gunnarif opened the pull request:
The test-data verifier needs more work but it will pull the latest data from the test-data repo and run through all the JSON files.

gabywh commented on dgc-testdata-tests/DGCTestDataVerifier.cs:
```diff
@@ -0,0 +1,84 @@
+                    //}
+                }
+                catch (Exception)
+                {
```
If the intention really is to just skip exceptions here, then a comment stating why would be useful - or is the error handling just missing ATM?

gabywh commented on dgc-testdata-tests/Program.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        static void Main(string[] args)
+        {
+            DGCTestDataVerifier.TestAll();
```
you're not `await`ing on async call - main could terminate before complete
GitHub Actions also mentions this:

> GitHub Actions / build
> dgc-testdata-tests/Program.cs#L13
> Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call.
>

gabywh commented on dgc-testdata-tests/SecretariatService.cs:
```diff
@@ -0,0 +1,29 @@
+            }
+            else
+            {
+                return Task.FromResult(new List<X509Certificate>().AsEnumerable());
```
not entirely sure of the purpose of this `Task.FromResult` call, but maybe that's just me - could you explain?

gunnarif replied:
just for testing purposes, but in practice this will be a async method but it is not the console testing app

gabywh commented on dgc-testdata-tests/SecretariatService.cs:
```diff
@@ -0,0 +1,29 @@
+
+        public void AddPublicKey(string keyId, X509Certificate cert)
+        {
+            trustedPublicKeys.Add(keyId, new List<X509Certificate> { cert });
```
A new list for each `cert` - or did you mean to add the cert to an existing `List<>` ?

gunnarif replied:
yes, as keyid are not necessarily unique in the gateway. But for testing purposes, I do not expect this but I need to comply with the interface.

gabywh commented on dgc.tests/UtilEncoderTests.cs:
```diff
@@ -0,0 +1,27 @@
+        [TestMethod]
+        public void MrzEncoder_TEST()
+        {
+            Assert.AreEqual("THETTA<ER<AD<PROFA", MrzEncoder.Encode("Þetta er að prófa"));
```
I have to admit that Icelandic isn't my strong point, I'll believe you on this one :)

gunnarif replied:
en: This is to test
:)

gabywh commented on dgc.tests/UtilEncoderTests.cs:
```diff
@@ -0,0 +1,27 @@
+        [TestMethod]
+        public void LuahModN_TEST()
+        {
+            Assert.AreEqual('E', LuhnModN.GenerateCheckCharacter("01LUX/1873751242292 3"));
```
see https://github.com/ehn-digital-green-development/ehn-dgc-schema/wiki/FAQ#uvci-checksum for final format

gabywh commented on dgc/DgCertificate.cs:
```diff
@@ -273,4 +280,20 @@ public partial class VaccinationEntry
+    {
+        public CustomDateTimeConverter()
+        {
+            DateTimeFormat = "yyyy-MM-dd";
```
see https://github.com/ehn-digital-green-development/ehn-dgc-schema/releases/tag/1.1.0 for up-coming, but yes I do appreciate that full date `YYYY-MM-DD` is v1.0 compliant.

gunnarif replied:
yes, I know. haven't had time to update to 1.2, expect to go to 1.3 when I have time.

gabywh commented on dgc/Gateway/GatewayService.cs:
```diff
@@ -0,0 +1,72 @@
+        private readonly X509Certificate2 _certificate;
+        
+        private HttpClient _httpClient;
+        private HttpClient HttpClient
```
If I am reading this correctly: readonly property of same name as `System.Net.Http.HttpClient`, could get confusing but syntactically ok I guess, lazy evaluation so only gets created at first point of calling.

I see you have some async code in here - how do you ensure that there are no race conditions on the check on line 26 `(if _httpClient != null) ... ` ?

gabywh commented on dgc/GreenCertificateEncoder.cs:
```diff
@@ -47,13 +83,13 @@ private byte[] GetCOSEBytes(byte[] cborBytes)
             msg.Content = cborBytes;
 
-            if (keypair.Private is Org.BouncyCastle.Crypto.Parameters.RsaPrivateCrtKeyParameters)
+            if (_keypair.Private is Org.BouncyCastle.Crypto.Parameters.RsaPrivateCrtKeyParameters)
```
Is one way of doing it - and certainly one I see a lot.
Q from my side: couldn't you provide an overloaded / generic func taking differing types so the type comparison is done by the compiler at compile time rather than a hand-coded if at runtime?

gabywh commented on dgc/dgc.csproj:
```diff
@@ -2,13 +2,37 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <EmbeddedResource Include="Valuesets\disease-agent-targeted.json" />
```
You would probably not want to embed these as resources but pull from known, good end-point (DCCG, when they start hosting it) at a point in time suitable to the program

gunnarif replied:
for now this will be like this, but when I can start getting the valueset from the gateway I will update.

gabywh reviewed (commented):
Looks good :)
Some comments for you to consider.

License: Apache v2.0 or EUPL tends to be used - for the ehn-dgc-schema I used Apache 2.0, - but of course this is your work so you are entitled to choose whichever license you feel is appropriate. I did see another comment about making the whole org ehn-digital-green-development Apache 2.0 - which is an option but that would remove original authors' rights. Your thoughts?

gunnarif commented:
@gabywh  thanks for the review
I did some fixes, updated the readme, added Apache 2.0 licence
Did some significant refactoring by removing the BouncyCastle library because of a bad support for HSMs

ryanbnl commented on dgc/MrzEncoder.cs:
```diff
@@ -3,34 +3,227 @@
-namespace dgc
+namespace DCC
 {
     public static class MrzEncoder
```
I would personally have a Transliteration interface, and an MRZ/MRTD implementation. Linking to the specs (doc 9303) would be nice too. For a simple example, maybe overkill, this is a really nice module so I can see it being something that people use / look at.

gabywh replied:
Transliteration interface would make sense if we know there are more transliterations coming along, but AFAIK there aren't - only the ICAO 9303 and only for very specific fields

gunnarif replied:
Currently, this is the only implementation and you need to reference it specifically when creating the JSON, not used by any other class, so I think it would be confusing to have this as an interface.

ryanbnl commented on dgc/Sign1CoseMessage.cs:
```diff
@@ -89,181 +85,95 @@ public byte[] EncodeToBytes()
             var protectedMap = CBORObject.NewMap();
-            if (alg == DGCertSupportedAlgorithm.ES256)
+
+            if (key is ECDsa)
```
switch() :-)

gabywh replied:
`switch` is one option - cleaner would be factory (one switch only) and sub-classing where the key type-specific operations are defined

ryanbnl replied:
Yup, subclassing is good too, cleaner but a larger change.

gunnarif replied:
Personally, I find using if statement more readable and cleaner than switch statements if the cases are few.
I really do not like subclassing.

ryanbnl commented on dgc/Sign1CoseMessage.cs:
```diff
@@ -89,181 +85,95 @@ public byte[] EncodeToBytes()
 
-            var signature = signer.GenerateSignature();
-            if (signLength > 0)
+            byte[] signature;
```
switch() :-)

ryanbnl commented on dgc/GreenCertificateVerifier.cs:
```diff
@@ -17,28 +18,30 @@ public GreenCertificateVerifier(SecretariatService secretariatService)
+            if (!verifyClock.HasValue) verifyClock = DateTime.Now;
+
+            var certificates = await _secretariatService.GetCertificate(cwt.CoseMessage.KID);
+            if (certificates.Any())
```
Multiple enumeration; usually I'd have the service return a collection type (i.e. readonly list) here, then you get a enumerator each time :)

gabywh replied:
> Multiple enumeration; 

In what way? `Any()` just checks to see if there is anything in the collection, doesn't really enumerate in the sense of going through the entire collection: ["The enumeration of source is stopped as soon as the result can be determined."](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=net-5.0) - which really should just check if there is an initial element or not and not even try to enumerate. Given the disappointing code I've seen when using ILDASM (even recently) on C# / .NET 5x, I could well imagine it still at least starts to enumerate rather than just directly check for presence of first element.

That said, I don't see why you need the call to `Any()` as the `foreach()` would simply nop if the collection empty - but in any case, this discussion of calling `Any()` is - at least AFAICS - more an optimization point rather than a program correctness point

> usually I'd have the service return a collection type (i.e. readonly list) here, then you get a enumerator each time :)

isn't that what `GetCertificate()` does? (which should really be called `GetCertificates()` ) - or have I just misread / assumed something that isn't there?

ryanbnl replied:
I should have said "potential multiple enumeration" - in the code base as it is now it is not a bug, but a code smell. 

The interface being called exposes IEnumerable<>, the contract that this defines is "enumerate this". This can lead to unexpected behaviour - in most cases when people expose IEnumerable directly via an ORM (and something that you - sadly - see way to often).

Despite never having encountered the bug in my own code (I know what I'm doing), I've seen it so often that I my code analysis tools are configured to treat this as an actual error :-)

gabywh replied:
> I should have said "potential multiple enumeration" - in the code base as it is now it is not a bug, but a code smell.
> 

? If the .net framework internally implements `Any()` as an enumeration process returning a boolean or does sth more sensible is nothing the caller of the framework method `Any()` can decide. As a caller, you can of course choose not to call `Any()` ;) Any code smell would therefore also be in the .NET framework rather than end-user code.

> The interface being called

well you would be calling an implementation of an interface, an interface by definition is not directly callable (abstract). So you would be calling something that implements the interface IEnumerable, in this case a Collection class.

> exposes IEnumerable<>, the contract that this defines is "enumerate this".

Here the return value is an object of type "Collection" (unless I'm mis-reading the code?) which amongst other things implements the interface `IEnumerable`, but [IEnumerable](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-5.0) itself does not have an 'Any()' method. The 'Any()' method is specified in [linq](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=net-5.0) and takes an `IEnumerable` argument, so yes the linq impl of `Any()` would be constrained to calling one of `IEnumerable.Current / MoveNext / Reset` but as the documentation for `Any()` states (ref in earlier comment) `Any()` will return as soon as it can determine the return value, so at least it doesn't run through the whole collection here, which was my initial point.

> This can lead to unexpected behaviour - in most cases when people expose IEnumerable directly via an ORM (and something that you - sadly - see way to often).

... but I don't see that is the case here - or am I missing sth?

In any case, the IEnumerable implementation called here is exposed from the .NET framework implementation for the Collection classes so no user-defined code here either, just calls to framework. Same for the (possibly unnecessary - separate issue) call to `Any()`. Not unknown for the .NET framework to have bugs in it, but that's a separate issue again.

ryanbnl replied:
Nope you're not missing anything - it's a "code cleanness" / pedantic code analysis point (so nothing will go wrong in the current implementation):

The interface returns an IEnumerable<X509Certificate2>:

`Task<IEnumerable<X509Certificate2>> GetCertificate(string kid);`

It's like declaring an interface that can return `Fruit`, when you really only want to see `CitrusFruit` with the default implementations doing just that. Fine, it will work, at least until some clever person adds a new implementation that returns `Apples`.

So, in this case, there's not a *need* to change anything, but doing so does improve the code.

gabywh replied:
> It's like declaring an interface that can return Fruit, when you really only want to see CitrusFruit

kind-of, but a little different: it's the generic type binding to IEnumerable (`X509Certificate2`) which makes it `CitrusFruit` -> but then here you have a very specific type so returning e.g. `IEnumerable<MyOtherType>` should be caught by the compiler (unless there is some clear inheritance relationship between `X509Certificate2` and `MyOtherType` - LSP, in effect).

gunnarif replied:
I changed this to IReadOnlyList

ryanbnl commented on dgc/DgCertificate.cs:
```diff
@@ -1,23 +1,45 @@
+        [JsonIgnore]
+        public DateTime? DateOfBirth 
+        { 
+            get {
```
formatting

ryanbnl commented on dgc/DgCertificate.cs:
```diff
@@ -1,23 +1,45 @@
         /// </summary>
         [JsonProperty("dob")]
-        public DateTimeOffset DateOfBirth { get; set; }
+        public string DateOfBirthString { get; set; }
```
I'm pretty sure that you can do this with just the JSON parser and a Nullable<DateTime>.

gabywh replied:
> I'm pretty sure that you can do this with just the JSON parser

Is one possibility - but if it works as-is, also ok I guess?

> and a Nullable

Why Nullable?

gunnarif replied:
Need to support partial dates that are for example only the year like "2000", sting type is necessary the have access to the real dob value in the QR code for display.

ryanbnl commented on dgc/DGCertSupportedAlgorithm.cs:
```diff
@@ -1,4 +1,4 @@
-﻿namespace DGC
+﻿namespace DCC
```
How are you handling multiple schema versions?

gunnarif replied:
currently, I don't see any need for it because a v1.3 is backwards compatible with v1.0

ryanbnl reviewed (commented):
Thanks for your contribution gunnarif. I've made a few comments in the code, I've kept myself focused to things that can affect correctness or understanding.

General:

- Copyright notices on the top of the files (see Base45Encodings.cs, you should add that for your own code.

- I'd recommend using dependency injection; you'll end up adding a number of interfaces but you'll make your code more flexible. If you fancy / are interested in it of course :-)

Let me know if you have any questions :)

gabywh commented on README.md:
```diff
@@ -1,9 +1,86 @@
+    Name = new Nam
+    {
+        FamilyName = lastName,
+        GivenName = firstName,
+        FamilyNameTransliterated = MrzEncoder.Encode(lastName),
+        GivenNameTraslitaerated = MrzEncoder.Encode(firstName),
```
I appreciate this is how it was named in the previous DCC versions. This was updated in 1.3.0 to Surname instead of Family Name to match the EU regs. JSON field name stayed the same though (perhaps confusingly, but at least keeping the fieldname the same didn't break all the existing code)

gabywh commented on README.md:
```diff
@@ -1,9 +1,86 @@
+X509Certificate2 cert = new X509Certificate2([ByteArrayOfTheCert])
+
+// Create CWT
+var ci = "URN:UVCI:01:IS:SomeUniqeID";
```
typo: SomeUniqueID

gabywh commented on README.md:
```diff
@@ -1,9 +1,86 @@
+        }
+    }
+};
+cwt.ExpiarationTime = testDate.AddHours(72);
```
typo: Expiaration -> Expiration

gabywh commented on README.md:
````diff
@@ -1,9 +1,86 @@
+```
+
+##Outstanding issues
+Verifying extended key parameters and the signing country has not been implented
````
typo: implemented

gabywh commented on README.md:
```diff
@@ -1,9 +1,86 @@
+
+##Outstanding issues
+Verifying extended key parameters and the signing country has not been implented
+Signature certificates when downloaded are not verified agains CA
```
typo: against

gabywh commented on README.md:
````diff
@@ -1,9 +1,86 @@
-It's meant to be a library to be used within other dotnet projects
+This is a dotnet standard library.
+
+## How to use
+### Verifing a certificate
+```c#
+// First connect to the DCC Gateway
````
I just highlighted a few lines for commenting purposes, but I really like this whole section, nice :)

gabywh commented on dgc-testdata-tests/DGCTestDataVerifier.cs:
```diff
@@ -0,0 +1,81 @@
+﻿using DCC;
+using DGC;
```
For DCC 1.3.0 you would want to unify it all under DCC and drop DGC

gabywh commented on dgc-testdata-tests/DGCTestDataVerifier.cs:
```diff
@@ -0,0 +1,81 @@
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace dgc.testdata.tests
+{
+    public class DGCTestDataVerifier
+    {
```
Similarly namespace + class names which should be just some ReSharper re-factor work AFAICS

gabywh commented on dgc-testdata-tests/DGCTestDataVerifier.cs:
```diff
@@ -0,0 +1,81 @@
+                }
+            }
+
+            var dcgDecoder = new GreenCertificateDecoder();
```
Names: 
1. Green -> Covid
1. **dcg**Decoder - typo?

gabywh commented on dgc-testdata-tests/DGCTestDataVerifier.cs:
```diff
@@ -0,0 +1,81 @@
+
+            var dcgDecoder = new GreenCertificateDecoder();
+            var secretariat = new SecretariatService();
+            var verifier = new GreenCertificateVerifier(secretariat);
```
name

gabywh commented on dgc-testdata-tests/DGCTestDataVerifier.cs:
```diff
@@ -0,0 +1,81 @@
+            foreach (var testdataAndFile in testdataset)
+            {
+                var testdata = testdataAndFile.Item1;
+                var errorMessage = "Test Passes";
```
minor: probably set errorMessage - if at all - to Test Fails and only set to Test Passes if all is known to be good?

gabywh commented on dgc-testdata-tests/Program.cs:
```diff
@@ -0,0 +1,16 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace dgc.testdata.tests
```
name of namespace

gabywh commented on dgc-testdata-tests/Program.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        static async Task Main(string[] args)
+        {
+            await DGCTestDataVerifier.TestAll();
```
name - won't list seperately from now on, will assume you just do a ReSharper name refactor or equivalant

gabywh commented on dgc/DgCertificate.cs:
```diff
@@ -1,23 +1,45 @@
+                } 
+                else if (int.TryParse(DateOfBirthString, out int year))
+                {
+                    return new DateTime(year, 1, 1);
```
Why 1,1 args here?

gunnarif replied:
so it will be a valid date. so I dob is "2000" then the dateofbirth field will be 2000.1.1.
This may differ between countries but dob is always a date in Iceland and if the day is not known then it is the first of that month.

gabywh commented on dgc/GreenCertificateVerifier.cs:
```diff
@@ -17,28 +18,30 @@ public GreenCertificateVerifier(SecretariatService secretariatService)
-            if (publicKeys.Any())
+            if (!verifyClock.HasValue) verifyClock = DateTime.Now;
+
+            var certificates = await _secretariatService.GetCertificate(cwt.CoseMessage.KID);
```
`GetCertificates(...)` ?

gunnarif replied:
changed this to count>0 because using IReadOnlyList

gabywh commented on dgc/Valuesets/disease-agent-targeted.json:
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "valueSetId": "disease-agent-targeted",
+  "valueSetDate": "2021-04-27",
+  "valueSetValues": {
+    "840539006": {
+      "display": "COVID-19",
+      "lang": "en",
+      "active": true,
+      "version": "http://snomed.info/sct/900000000000207008/version/20210131",
+      "system": "http://snomed.info/sct"
+    }
+  }
+}
```
Just a comment: I take it you do know the valuesets should be distributed by the DCCG? If so, then ok with these here as just a snapshot for development purposes only

gabywh reviewed (commented):
Added my comments: nothing I would absolutely insist on before approving this - but you may wish to take a look at my comments first and decide for yourself what you would like to improve out of them (maybe even all of them?).

gunnarif commented:
@gabywh and @ryanbnl thanks for all the good comment.
I tried to resolve most of them, but some I left unchanged. 

I think many countries are using the implementation and this is almost a two month old pull request that fixed many issues, for example the time in dob was fixed 7. may.
From an implementation standpoint, I have not found any functional issues.
Is it possible to accept this pull request, any new issue will then become a new pull request?

ryanbnl commented:
Yeah that's good :)
